File: Boj_py/1260.py

N, M, V = map(int, input().split())

graph = {}

for i in range(1, N+1):
    graph[i] = []

for i in range(0, M):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)


def dfs(graph, start):
    visited = []
    stack = [start]
    while stack:
        v = stack.pop()
        if v not in visited:
            visited.append(v)
            print(v, end=" ")
            for w in graph[v]:
                stack.append(w)


for i in graph.keys():
    graph[i].sort(reverse=True)  # sort()랑 reverse()다르다!

dfs(graph, V)
print()


def bfs(graph, start):
    visited = [start]
    queue = [start]
    while queue:
        v = queue.pop(0)
        for w in graph[v]:
            if w not in visited:
                visited.append(w)
                print(w, end=" ")
                queue.append(w)


for i in graph.keys():
    graph[i].sort()


print(V, end=" ")
bfs(graph, V)

File: Boj_kt/DFS_BFS/1260.kt

fun main() {
    val br = System.`in`.bufferedReader()
    val (n, m, v) = readln().split(" ").map { it.toInt() }
    val graph = Array(n + 1) { ArrayList<Int>() }

    val visitedDfs = BooleanArray(n + 1) { false }
    val visitedBfs = BooleanArray(n + 1) { false }

    repeat(m) {
        val (v1, v2) = br.readLine().split(" ").map { it.toInt() }
        graph[v1].add(v2)
        graph[v2].add(v1)
    }
    br.close()

    graph.forEach {
        it.sort()
    }


    dfs(graph, v, visitedDfs)
    println()
    bfs(graph, v, visitedBfs)
}

fun dfs(graph: Array<ArrayList<Int>>, v: Int, visited: BooleanArray) {
    visited[v] = true
    print("$v ")
    graph[v].forEach {
        if (!visited[it]) {
            dfs(graph, it, visited)
        }
    }
}

fun bfs(graph: Array<ArrayList<Int>>, v: Int, visited: BooleanArray) {
    val q = ArrayDeque<Int>()
    q.addFirst(v)
    visited[v] = true
    while (q.isNotEmpty()) {
        val x = q.removeLast()
        print("$x ")
        graph[x].forEach {
            if (!visited[it]) {
                q.addFirst(it)
                visited[it] = true
            }
        }
    }
}